### PR TITLE
More metrics for the EL monitor

### PR DESCRIPTION
**Pull request by zah**
* `engine_api_response_time` provides a histogram for the Engine API response times for each unique pair ot URL and request type.

* All engine API requests are now tracked

Other changes:

The client will no longer exit on start-up if it fails to connect to a properly configured EL node.

### CODE REVIEW

**Comment by jakubgs** [comment]
Looking good:
```
admin@linux-02.he-eu-hel1.nimbus.prater:~ % c http://10.14.0.70:9202/metrics | grep engine_api_response_time | grep newPayload
engine_api_response_time_sum{url="http://localhost:8554",request="newPayload"} 6.565
engine_api_response_time_count{url="http://localhost:8554",request="newPayload"} 65.0
engine_api_response_time_created{url="http://localhost:8554",request="newPayload"} 1678377338.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="0.1"} 36.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="0.25"} 65.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="0.5"} 65.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="1.0"} 65.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="2.0"} 65.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="5.0"} 65.0
engine_api_response_time_bucket{url="http://localhost:8554",request="newPayload",le="+Inf"} 65.0
```

**Comment by jakubgs** [comment]
Seems to work, but I'm not getting values on the vertical scale.Weird.

image

File: beacon_chain/eth1/eth1_monitor.nim [hunk 1]
```diff
@@ -259,6 +259,11 @@ declareCounter engine_api_responses,
   "Number of successful requests to the newPayload Engine API end-point",
   labels = ["url", "request", "status"]
 
+declareHistogram engine_api_response_time,
```
Comment (jakubgs) [inline beacon_chain/eth1/eth1_monitor.nim:262]: A more conventional naming for histograms is `duration`, not `time`. Also the name should include units used.
So maybe more like `engine_api_request_duration_seconds`.

File: beacon_chain/eth1/eth1_monitor.nim [hunk 2]
```diff
@@ -259,6 +259,11 @@ declareCounter engine_api_responses,
...
+  "Time(s) used to generate signature usign remote signer",
+   buckets = [0.1, 0.25, 0.5, 1.0, 2.0, 5.0],
```
Comment (jakubgs) [inline beacon_chain/eth1/eth1_monitor.nim:264]: Could we maybe have a bit more fine grained buckets? Here's an example from [Cortex](https://cortexmetrics.io/):
```
admin@master-01.do-ams3.metrics.hq:~ % c 0:9092/metrics | grep cortex_request_duration_seconds_bucket | grep ingester_ring
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.005"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.01"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.025"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.05"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.1"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.25"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="0.5"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="1"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="2.5"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="5"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="10"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="25"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="50"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="100"} 2
cortex_request_duration_seconds_bucket{method="GET",route="ingester_ring",status_code="200",ws="false",le="+Inf"} 2
```

File: beacon_chain/eth1/eth1_monitor.nim (hunk 1 above)
Comment (jakubgs) [reply beacon_chain/eth1/eth1_monitor.nim:262]: >...should have a suffix describing the unit, in plural form. Note that an accumulating count has total as a suffix, in addition to the unit if applicable.
>- `http_request_duration_seconds`
>- `node_memory_usage_bytes`

https://prometheus.io/docs/practices/naming/#metric-names

**Comment by jakubgs** [comment]
Ok, I see what was wrong. The `Legend` has to show only the `le`:

image

And now it looks correct:

image

But yeah, I'd appreciate some more granular buckets.

**Comment by zah** [comment]
The suggested changes have been pushed to `unstable` in e808fda02cccc06a7b3a413f44371dd17c682279. The new metric name is `engine_api_request_duration_seconds`

**Comment by jakubgs** [comment]
Broken tho:
```
/data/beacon-node-prater-unstable/repo/beacon_chain/eth1/eth1_monitor.nim(304, 7) Error: undeclared identifier: 'engine_api_response_time'
candidates (edit distance, scope distance); see '--spellSuggest':
 (4, 6): 'engine_api_responses' [var declared in /data/beacon-node-prater-unstable/repo/beacon_chain/eth1/eth1_monitor.nim(258, 16)]
```

**Comment by jakubgs** [comment]
Pretty cool:

### `linux-04.he-eu-hel1.nimbus.prater`

image

### `linux-05.he-eu-hel1.nimbus.prater`

image

### `linux-06.he-eu-hel1.nimbus.prater`

image

---

Here we can clearly see that the more validators you have the more strained the EL node is.

**Comment by jakubgs** [comment]
Interesting jump in `getBlockByNumber`:

image

https://metrics.status.im/d/pgeNfj2Wz23/nimbus-fleet-testnets?orgId=1&refresh=5m&var-instance=linux-05.he-eu-hel1.nimbus.prater&var-container=beacon-node-prater-unstable&from=1678394472978&to=1678405272978

**Comment by zah** [comment]
That's expected. The earlier version of the code didn't track all requests.

**Pull request merged**